pallets/msa/src/lib.rs

```diff
@@ -1655,6 +1643,14 @@ impl<T: Config + Send + Sync> CheckFreeExtrinsicUse<T> {
 			key_count == 1,
 			InvalidTransaction::Custom(ValidityError::InvalidMoreThanOneKeyExists as u8)
 		);
+
+		let num_delegations =
+			DelegatorAndProviderToDelegation::<T>::iter_key_prefix(DelegatorId(msa_id)).count();
+		ensure!(
+			num_delegations == 0,
+			InvalidTransaction::Custom(ValidityError::InvalidNonZeroProviderDelegations as u8)
+		);
```

Comment:
I don't think we need the whole count here, just that any exist.

```suggestion
		let has_delegations: bool =
			DelegatorAndProviderToDelegation::<T>::iter_key_prefix(DelegatorId(msa_id)).any(|_| true);

		ensure!(
			!has_delegations,
			InvalidTransaction::Custom(ValidityError::InvalidNonZeroProviderDelegations as u8)
		);
```